### PR TITLE
[MSHADE-36] Use reduced dependencies

### PR DESCRIPTION
**Pull request by nielsbasjes**
This is a proposed fix for the issue https://issues.apache.org/jira/browse/MSHADE-36 which has been open for over 11 years.

The implementation is activated when the setting "useDependencyReducedPomInJar" is set to true.
It then essentially injects two ResourceTransformers from code
1) Remove the original "META-INF/maven/${project.groupId}/${project.artifactId}/pom.xml".
2) Include the dependency-reduced-pom.xml under the name of the original pom.xml into the final jar file.

This includes an integration test to verify this actually happens.

----------------

Following this checklist to help us incorporate your 
contribution quickly and easily:

 - [x] Make sure there is a [JIRA issue](https://issues.apache.org/jira/browse/MSHADE) filed 
       for the change (usually before you start working on it).  
      https://issues.apache.org/jira/browse/MSHADE-36
 - [x] Each commit in the pull request should have a meaningful subject line and body.
 - [x] Format the pull request title like `[MSHADE-XXX] - Fixes bug in ApproximateQuantiles`,
       where you replace `MSHADE-XXX` with the appropriate JIRA issue. Best practice
       is to use the JIRA issue title in the pull request title and in the first line of the 
       commit message.
 - [x] Write a pull request description that is detailed enough to understand what the pull request does, how, and why.
 - [x] Run `mvn clean verify` to make sure basic checks pass. A more thorough check will 
       be performed on your pull request automatically.
 - [x] You have run the integration tests successfully (`mvn -Prun-its clean verify`).

If your pull request is about ~20 lines of code you don't need to sign an
[Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf) if you are unsure please ask on the developers list.

To make clear that you license your contribution under 
the [Apache License Version 2.0, January 2004](http://www.apache.org/licenses/LICENSE-2.0)
you have to acknowledge this by using the following check-box.

 - [x] I hereby declare this contribution to be licensed under the [Apache License Version 2.0, January 2004](http://www.apache.org/licenses/LICENSE-2.0)

 - [x] In any other case, please file an [Apache Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf). I'm a committer and PMC for Apache Avro.

### CODE REVIEW

**Comment by nielsbasjes** [comment]
I noticed that there will be a conflict with https://github.com/apache/maven-shade-plugin/pull/22 because I moved the creation of the dpr to a slightly different place (more forward instead of more backward).

File: src/it/MSHADE-36-inject-dep-reduced-pom-in-final/verify.bsh
Comment (Tibor17) [inline src/it/MSHADE-36-inject-dep-reduced-pom-in-final/verify.bsh]: Pls call this in `finally {}` block.
It gives at least some guarantee that we do not have open streams when deleting workspace on Jenkins, thx.

Comment (nielsbasjes) [reply src/it/MSHADE-36-inject-dep-reduced-pom-in-final/verify.bsh]: I've updated the script. Please check if this is what you meant.

**Comment by struberg** [comment]
What is the status of this ticket? @Tibor17 was there something missing still?

**Comment by rfscholte** [comment]
I talked some time ago with @hboutemy about shading and poms. Which is the right pom? Think about licenses, etc. I can image we need a `shaded` scope, which keeps all original information and respects the shading-aspects. But that is another topic.

**Comment by nielsbasjes** [comment]
@elharo I would like to see this problem fixed. If there is anything I need to change then I'm happy to do that. Just say what the desired solution is.

**Comment by elharo** [comment]
Looks like for this issue to move forward it needs a proposal (not a PR) that can be discussed on the dev mailing list. If you can get consensus from the project maintainers on what a fix should look like--i.e. what would go into a project POM and what the output would be--then it could be implemented. But right now it doesn't look like there's consensus on the form of the solution.

**Comment by felfert** [comment]
> Looks like for this issue to move forward it needs a proposal (not a PR) that can be discussed on the dev mailing list.

Perhaps, there could be a pragmatic solution to this: Instead of directly using the dependency-reduced pom directly, there
could be an option to use a user-supplied pom instead. Then the user could copy the generated pom, edit it (adding licence info etc) and then in a second run, that pom could be used in the jar?

**Comment by Snipx** [comment]
Dear Maintainers/Contributors,
I think this is a terribly useful feature - is there any chance it can be given a bit more priority?

File: src/it/MSHADE-36-inject-dep-reduced-pom-in-final/invoker.properties
Comment (michael-o) [inline src/it/MSHADE-36-inject-dep-reduced-pom-in-final/invoker.properties:19]: Why is that necessary?

File: src/it/MSHADE-36-inject-dep-reduced-pom-in-final/pom.xml
Comment (michael-o) [inline src/it/MSHADE-36-inject-dep-reduced-pom-in-final/pom.xml:44]: These should match the plugin's Java requirement.

File: src/it/MSHADE-36-inject-dep-reduced-pom-in-final/pom.xml
Comment (michael-o) [inline src/it/MSHADE-36-inject-dep-reduced-pom-in-final/pom.xml:39]: normal => original

File: src/it/MSHADE-36-inject-dep-reduced-pom-in-final/src/main/java/nl/example/Main.java
Comment (michael-o) [inline src/it/MSHADE-36-inject-dep-reduced-pom-in-final/src/main/java/nl/example/Main.java:20]: Use `com.example`

File: src/main/java/org/apache/maven/plugins/shade/mojo/ShadeMojo.java
Comment (michael-o) [inline src/main/java/org/apache/maven/plugins/shade/mojo/ShadeMojo.java:297]: Update this one

File: src/main/java/org/apache/maven/plugins/shade/mojo/ShadeMojo.java
Comment (michael-o) [inline src/main/java/org/apache/maven/plugins/shade/mojo/ShadeMojo.java:450]: You have moved this before the Shader works. No changes in behavior?

File: src/main/java/org/apache/maven/plugins/shade/resource/UseDependencyReducedPomXml.java
Comment (michael-o) [inline src/main/java/org/apache/maven/plugins/shade/resource/UseDependencyReducedPomXml.java:32]: Please name this `UseDependencyReducedPom`

**Review by michael-o — CHANGES_REQUESTED** [review]

**Comment by michael-o** [comment]
I am inclined to merge this after a few small changes since this is optional and does not affect default behavior. Respecing licenses and other attributions is solely a dev's task. Not ours.

**Comment by michael-o** [comment]
@nielsbasjes Please also rebase on top of master to resolve merge conflicts.

**Comment by felfert** [comment]
Wow. Finally some progress.
@michael-o Please would you be so kind to have a look at #26 as well? Those two PR kind of go hand-in-hand and finally would make shading usable in multi-module projects. Thanks.

**Comment by michael-o** [comment]
@felfert As far as I can see #26 requires this PR, doesn't it? If so, I would like to complete this one and have a look at the other. I need to understand the problem first.

**Comment by felfert** [comment]
> @felfert As far as I can see #26 requires this PR, doesn't it? If so, I would like to complete this one and have a look at the other. I need to understand the problem first.

No, as far as I interpret the comments #26 (See the very first line in the initial text), #26 **includes** #25.
As for understanding the problem, maybe a simple use case shows the dilemma:

Have a look at https://github.com/jenkinsci/jclouds-plugin (A plugin which uses jclouds which in turn uses guava) and
https://github.com/felfert/jclouds-shaded (A shaded version of jclouds, where the guava dependency is shaded, because jenkins itself uses a different guava version). The latter exists for the sole purpose of shading.
Normally, I never would publish a separate project only to facilitate shading. Instead, I would make the shaded jclouds a submodule of the plugin. However maven currently forces me to do this, because even with a workaround for #25, maven still uses the unshaded pom during reactor build.

Cheers
 -Fritz

**Comment by michael-o** [comment]
> > @felfert As far as I can see #26 requires this PR, doesn't it? If so, I would like to complete this one and have a look at the other. I need to understand the problem first.
> 
> No, as far as I interpret the comments in #26 (See the very first line in the initial text), #26 **includes** #25. Also, if you click on the commits tab of #26 you see TWO commits, where one is identical to **this** PR.

This is a contradiction, isn't it? If #26 includes this PR then it *requires* it. I will first resolve this one waiting for @nielsbasjes  then continue with the rest.

**Comment by nielsbasjes** [comment]
I'm picking up the reviews and comments.
It has been a while and several things have changed.

**Comment by michael-o** [comment]
> 
> 
> I'm picking up the reviews and comments.
> It has been a while and several things have changed.

Thanks, looking forward too. If all goes well we'll have a release this month.

File: src/it/MSHADE-36-inject-dep-reduced-pom-in-final/invoker.properties
Comment (nielsbasjes) [reply src/it/MSHADE-36-inject-dep-reduced-pom-in-final/invoker.properties:19]: Don't remember. I removed the java.version thing and it worked.

File: src/it/MSHADE-36-inject-dep-reduced-pom-in-final/pom.xml
Comment (nielsbasjes) [reply src/it/MSHADE-36-inject-dep-reduced-pom-in-final/pom.xml:44]: Apparently I copied a bit too much.
I removed this and it works.

File: src/it/MSHADE-36-inject-dep-reduced-pom-in-final/src/main/java/nl/example/Main.java
Comment (nielsbasjes) [reply src/it/MSHADE-36-inject-dep-reduced-pom-in-final/src/main/java/nl/example/Main.java:20]: Done

File: src/main/java/org/apache/maven/plugins/shade/mojo/ShadeMojo.java
Comment (nielsbasjes) [reply src/main/java/org/apache/maven/plugins/shade/mojo/ShadeMojo.java:297]: I assume @since 3.2.5 ?

File: src/main/java/org/apache/maven/plugins/shade/resource/UseDependencyReducedPomXml.java
Comment (nielsbasjes) [reply src/main/java/org/apache/maven/plugins/shade/resource/UseDependencyReducedPomXml.java:32]: Done

File: src/main/java/org/apache/maven/plugins/shade/mojo/ShadeMojo.java
Comment (nielsbasjes) [reply src/main/java/org/apache/maven/plugins/shade/mojo/ShadeMojo.java:450]: With the tests present in the project today 2 fail with my changes in place.
So this does indeed need some rework.

File: src/main/java/org/apache/maven/plugins/shade/mojo/ShadeMojo.java
Comment (michael-o) [reply src/main/java/org/apache/maven/plugins/shade/mojo/ShadeMojo.java:297]: 3.3.0

File: src/main/java/org/apache/maven/plugins/shade/mojo/ShadeMojo.java
Comment (nielsbasjes) [reply src/main/java/org/apache/maven/plugins/shade/mojo/ShadeMojo.java:461]: Apparently the problem that failed the tests made two files. 
Fix that now all tests pass.

Still the question "No changes in behavior?" remains a valid one.

For now: Let's assume there IS a change the current tests did not pick up.
I'll look into it.

**Comment by nielsbasjes** [comment]
Regarding #26: That one indeed requires this fix to be in place first.
Once this one (#25) has been merged I'll update #26 for a renewed review round.

**Comment by michael-o** [comment]
@nielsbasjes Something is wrong, GitHub shows 97 commits to be merged. Did you rebase incorrectly?

File: src/main/java/org/apache/maven/plugins/shade/mojo/ShadeMojo.java
Comment (nielsbasjes) [reply src/main/java/org/apache/maven/plugins/shade/mojo/ShadeMojo.java:461]: My verification steps:

0. All integration tests pass.
1. Checkout both master and my version into separate directories.
2. Run in both: `MAVEN_OPTS=-Dinvoker.parallelThreads=2 mvn clean verify -Prun-its` under javac 11.0.11
3. Get this tool to compare the jars https://github.com/scala/jardiff
4. Run these magical incantations on my Linux machine:
`find target/it -type f -name '*.jar' | while read JAR ; do java -jar ../jardiff.jar ${JAR} ../maven-shade-plugin-master/${JAR} ; done > jar-changes.log`
`find target/it -type f -name '*pom*' | while read POM ; do diff  ${POM} ../maven-shade-plugin-master/${POM} ; done > pom-changes.log`
5. Manually go through the output.
Differences logged:
- Files like surefirebooter2116138492290729396.jar are unique (Ignore)
- The test files I created are only present in mine (Ignore)
- All files have a difference in the timestamps when the files were generated (pom.properties) (Ignore)
- Some tests include the plugin itself and show differences because of the new feature (Ignore)

@michael-o Is this sufficient verification?

**Comment by nielsbasjes** [comment]
> @nielsbasjes Something is wrong, GitHub shows 97 commits to be merged. Did you rebase incorrectly?

Yes, something went wrong here.
I'm fixing it.

File: src/main/java/org/apache/maven/plugins/shade/mojo/ShadeMojo.java
Comment (michael-o) [reply src/main/java/org/apache/maven/plugins/shade/mojo/ShadeMojo.java:461]: Will verify!

**Comment by michael-o** [comment]
@nielsbasjes Let me explicitly know when you are done from your PoV and I will review and test everything.

**Comment by nielsbasjes** [comment]
@michael-o I have fixed it. CI builds all pass now.
Pushing a squased version for cleaner merging.

File: src/it/MSHADE-36-inject-dep-reduced-pom-in-final/invoker.properties
Comment (michael-o) [inline src/it/MSHADE-36-inject-dep-reduced-pom-in-final/invoker.properties:19]: Still present

File: src/it/MSHADE-36-inject-dep-reduced-pom-in-final/pom.xml
Comment (michael-o) [inline src/it/MSHADE-36-inject-dep-reduced-pom-in-final/pom.xml:43]: Still present

File: src/it/MSHADE-36-inject-dep-reduced-pom-in-final/pom.xml
Comment (michael-o) [inline src/it/MSHADE-36-inject-dep-reduced-pom-in-final/pom.xml:85]: Should have been com.example...

File: src/it/MSHADE-36-inject-dep-reduced-pom-in-final/src/main/java/nl/example/Main.java
Comment (michael-o) [inline src/it/MSHADE-36-inject-dep-reduced-pom-in-final/src/main/java/nl/example/Main.java:20]: Same here

File: src/main/java/org/apache/maven/plugins/shade/mojo/ShadeMojo.java
Comment (michael-o) [inline src/main/java/org/apache/maven/plugins/shade/mojo/ShadeMojo.java:292]: 3.3.0

**Review by michael-o — CHANGES_REQUESTED** [review]

File: src/main/java/org/apache/maven/plugins/shade/mojo/ShadeMojo.java
Comment (nielsbasjes) [reply src/main/java/org/apache/maven/plugins/shade/mojo/ShadeMojo.java:292]: Apparently more went wrong at my end.
Fixing all of those.
Re checking all historical feedback.

**Comment by michael-o** [comment]
Going through now...

File: src/main/java/org/apache/maven/plugins/shade/mojo/ShadeMojo.java
Comment (michael-o) [inline src/main/java/org/apache/maven/plugins/shade/mojo/ShadeMojo.java:290]: This one is odd. If you don't use the DRP you use the JAR?. Shouldn't it read:
Add dependency reduced POM to the JAR instead of the original one provided by the project.

File: src/main/java/org/apache/maven/plugins/shade/mojo/ShadeMojo.java
Comment (michael-o) [inline src/main/java/org/apache/maven/plugins/shade/mojo/ShadeMojo.java:466]: Maybe this should be nested in the `if ( createDependencyReducedPom )` clause instead of the exception. The documentation can tell that this is a noop if `createDependencyReducedPom ` isn't set?! WDYT?

File: src/main/java/org/apache/maven/plugins/shade/resource/UseDependencyReducedPom.java
Comment (michael-o) [inline src/main/java/org/apache/maven/plugins/shade/resource/UseDependencyReducedPom.java:30]: it's => its (possive pronoun)

File: src/main/java/org/apache/maven/plugins/shade/resource/UseDependencyReducedPom.java
Comment (michael-o) [inline src/main/java/org/apache/maven/plugins/shade/resource/UseDependencyReducedPom.java:34]: Drop the `Xml` here and below.

**Review by michael-o — CHANGES_REQUESTED** [review]
A few issues

File: src/it/projects/MSHADE-36-inject-dep-reduced-pom-in-final/verify.bsh
Comment (nielsbasjes) [reply src/it/projects/MSHADE-36-inject-dep-reduced-pom-in-final/verify.bsh:34]: The minimizeJar is true and this is to check that that still works.

File: src/main/java/org/apache/maven/plugins/shade/mojo/ShadeMojo.java
Comment (nielsbasjes) [reply src/main/java/org/apache/maven/plugins/shade/mojo/ShadeMojo.java:290]: That is definitely a typo. Your text is a lot clearer.

File: src/main/java/org/apache/maven/plugins/shade/mojo/ShadeMojo.java
Comment (nielsbasjes) [reply src/main/java/org/apache/maven/plugins/shade/mojo/ShadeMojo.java:466]: I structured it to give an error in case of an invalid configuration (use but don't create) instead of silently doing a noop.

File: src/main/java/org/apache/maven/plugins/shade/resource/UseDependencyReducedPom.java
Comment (nielsbasjes) [reply src/main/java/org/apache/maven/plugins/shade/resource/UseDependencyReducedPom.java:30]: Yes

File: src/it/projects/MSHADE-36-inject-dep-reduced-pom-in-final/verify.bsh
Comment (michael-o) [reply src/it/projects/MSHADE-36-inject-dep-reduced-pom-in-final/verify.bsh:34]: My fault, a misunderstanding!

File: src/main/java/org/apache/maven/plugins/shade/mojo/ShadeMojo.java
Comment (michael-o) [reply src/main/java/org/apache/maven/plugins/shade/mojo/ShadeMojo.java:466]: @rfscholte In a few plugin we explicitly ignore params if a condition is not met. Should we apply here the same logic?

Comment (rfscholte) [reply src/main/java/org/apache/maven/plugins/shade/mojo/ShadeMojo.java:466]: I agree with @michael-o . It is like calling `clean` when there is nothing to clean. Not worth an exception.

Comment (nielsbasjes) [reply src/main/java/org/apache/maven/plugins/shade/mojo/ShadeMojo.java:466]: Ok, I'll change it.

Comment (nielsbasjes) [reply src/main/java/org/apache/maven/plugins/shade/mojo/ShadeMojo.java]: Done.

File: src/main/java/org/apache/maven/plugins/shade/mojo/ShadeMojo.java
Comment (michael-o) [inline src/main/java/org/apache/maven/plugins/shade/mojo/ShadeMojo.java:290]: The documentation should also say:
If {@code createDependencyReducedPom} is {@code false} this parameter will be ignored.

**Review by michael-o — CHANGES_REQUESTED** [review]

File: src/main/java/org/apache/maven/plugins/shade/mojo/ShadeMojo.java
Comment (nielsbasjes) [reply src/main/java/org/apache/maven/plugins/shade/mojo/ShadeMojo.java:290]: LOL, I was just about to push the same change :)

**Comment by michael-o** [comment]
Works for me on:
```
$ mvn -v
Apache Maven 3.5.4 (1edded0938998edf8bf061f1ceb3cfdeccf443fe; 2018-06-17T20:33:14+02:00)
Maven home: /usr/local/apache-maven-3.5.4
Java version: 1.8.0_292, vendor: OpenJDK BSD Porting Team, runtime: /usr/local/openjdk8/jre
Default locale: de_DE, platform encoding: UTF-8
OS name: "freebsd", version: "12.2-stable", arch: "amd64", family: "unix"
```

Waiting for ASF Jenkins to complete

**Comment by michael-o** [comment]
@nielsbasjes Please continue (rebase) the next PR and let me know when you are done.

**Comment by nielsbasjes** [comment]
Yes, will do

**Comment by Snipx** [comment]
Excellent cooperation, thank you folks! So glad to see this fixed

**Comment by michael-o** [comment]
> 
> 
> Excellent cooperation, thank you folks! So glad to see this fixed

Took way too long to process, but people are busy.

**Comment by jira-importer** [comment]
Resolve #404

**Comment by jira-importer** [comment]
Resolve #404

**Pull request closed without merging**
